src/Planners/src/UnicycleTrajectoryGenerator.cpp

```diff
@@ -0,0 +1,844 @@
+    }
+
+    // the trajectory was not finished the new trajectory will be attached at the next merge point
+    else
+    {
+        // Searches for the first merge point that is at least m_plannerAdvanceTimeSteps steps away
+        auto firstMergePointAvailable
+            = std::find_if(m_pImpl->trajectory.mergePoints.begin(),
+                           m_pImpl->trajectory.mergePoints.end(),
+                           [this](size_t input) {
+                               return input >= this->m_pImpl->parameters.plannerAdvanceTimeSteps;
+                           });
+
+        if (firstMergePointAvailable != m_pImpl->trajectory.mergePoints.end())
+        {
+            if (m_pImpl->newTrajectoryRequired)
+                return true;
+
+            m_pImpl->newTrajectoryMergeCounter = *firstMergePointAvailable;
+        } else
+        {
+            if (m_pImpl->newTrajectoryRequired)
+                return true;
+
+            m_pImpl->newTrajectoryMergeCounter = m_pImpl->parameters.plannerAdvanceTimeSteps;
+        }
+    }
+
+    m_pImpl->newTrajectoryRequired = true;
+
+    return true;
+}
+
+bool Planners::UnicycleTrajectoryGenerator::advance()
+{
+    constexpr auto logPrefix = "[UnicycleTrajectoryGenerator::advance]";
+
+    if (m_pImpl->state == Impl::FSM::NotInitialized)
+    {
+        log()->error("{} The Unicycle Trajectory Generator has never been initialized.", logPrefix);
+        return false;
+    }
+
+    // if a new trajectory is required check if its the time to evaluate the new trajectory or
+    // the time to attach new one
+    if (m_pImpl->newTrajectoryRequired)
+    {
+        // when we are near to the merge point the new trajectory is evaluated
+        if (m_pImpl->newTrajectoryMergeCounter == m_pImpl->parameters.plannerAdvanceTimeSteps)
+        {
+
+            std::chrono::nanoseconds initTimeTrajectory
+                = m_pImpl->time + m_pImpl->newTrajectoryMergeCounter * m_pImpl->parameters.dt;
+
+            manif::SE3d measuredTransform = m_pImpl->trajectory.isLeftFootLastSwinging.front()
+                                                ? m_pImpl->trajectory.rightFootTransform.at(
+                                                    m_pImpl->newTrajectoryMergeCounter)
+                                                : m_pImpl->trajectory.leftFootTransform.at(
+                                                    m_pImpl->newTrajectoryMergeCounter);
+
+            // ask for a new trajectory (and spawn an asynchronous thread to compute it)
+            m_pImpl->mutex.lock();
+            if (m_pImpl->unicyclePlannerState == Impl::unicycleTrajectoryPlannerState::Running)
+            {
+                log()->error("{} The unicycle planner is still running.", logPrefix);
+                return false;
+            }
+            m_pImpl->unicyclePlannerState = Impl::unicycleTrajectoryPlannerState::Called;
+            m_pImpl->mutex.unlock();
```

Comment:
Rather than ``mutex.lock()``/``mutex.unlock()`` it is better to use a RAII paradigm. See https://kayari.org/cxx/antipatterns.html#locking-mutex

Reply:
I'll check this out, thanks!

Reply:
I should have implemented it with https://github.com/ami-iit/bipedal-locomotion-framework/pull/845/commits/908d771468887f845e6429b5155544e5cb2c5ee4